### PR TITLE
DOC: Simd optimization documentation

### PR DESCRIPTION
**Pull request by mattip**
@seiko2plus I reformatted your document and made it into a separate PR here as rst.

### CODE REVIEW

**Review by rossbar — APPROVED** [review]
Thanks @seiko2plus and @mattip for putting this together - it's an interesting read!

**Review by seberg — COMMENTED** [review]
Thanks, mostly nitpicky stuff, so I think we can just fix it up slightly and merge, and then iterate.  The first part explaining how to build (which is read by most users) seems pretty clear to me.  When it comes to the implementation side and infrastructure it was much less clear to me to be honest.

If it helps a lot, I can try to make one type-edit style pass on it, similar to what Ross did.  I had started before ross did his pass, so its possible some of the things do not apply anymore.

File: doc/source/reference/simd/simd-optimizations.rst [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+       ``IBM/POWER`` ``big-endian mode``      ``NONE``
+       ``IBM/POWER`` ``little-endian mode``   ``VSX`` ``VSX2``
+       ``ARMHF``                              ``NONE``
+       ``ARM64`` ``AARCH64``                  ``NEON`` ``NEON_FP16`` ``NEON_VFPV4``
+                                              ``ASIMD``
+      ======================================  =======================================
+
+- ``MAX``: Enables all supported CPU features by the Compiler and platform.
+
+- ``Operators-/+``: remove or add features, useful with options ``MAX``, ``MIN`` and ``NATIVE``.
```
[the remark this answers is not in the document]
Comment (seiko2plus) [reply doc/source/reference/simd/simd-optimizations.rst:75]: +1, the only example is explained within the context of the default value of `--cpu-dispatch`:
>  "max -xop -fma4" which enables all CPU features, except for AMD legacy features.

File: doc/source/reference/simd/simd-optimizations.rst [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+
+**Dispatch-able features** are our dispatched set of additional optimizations
+that were configured via ``--cpu-dispatch``. They are not activated by
+default and are always guarded by other C definitions prefixed with
+``NPY__CPU_TARGET_``. C definitions ``NPY__CPU_TARGET_`` are only
+enabled within **dispatch-able sources**.
+
+.. _dispatchable-sources:
+
+6- Dispatch-able sources and configuration statements
```
[the remark this answers is not in the document]
Comment (seiko2plus) [reply doc/source/reference/simd/simd-optimizations.rst]: +1,

> I assume this comes from arabic using more of a dash here

yep (^_^)

File: doc/source/reference/simd/simd-optimizations.rst [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+      // @param CB, a callback macro that expected to be called multiple times depending
+      // on the required optimizations, the callback should receive the following arguments:
+      //  1- The pending calls of @param CHK filled up with the required CPU features,
+      //     that need to be tested first in runtime before executing call belong to
+      //     the compiled object.
+      //  2- The required optimization name, same as in 'NPY__CPU_TARGET_CURRENT'
+      //  3- Extra arguments in the macro itself
+      //
+      // By default the callback calls are sorted depending on the highest interest
+      // unless the policy "$keep_sort" was in place within the configuration statements
```
[the remark this answers is not in the document]
Comment (seiko2plus) [reply doc/source/reference/simd/simd-optimizations.rst:411]: what about mixing the code with RST instead? can be applied?

File: doc/source/reference/simd/simd-optimizations.rst [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+      void simd_whoami_SSE41(const char *extra_info);
+
+      void trigger_me(void)
+      {
+          // bring the auto-gernreated config header
+          // which contains config macros 'NPY__CPU_DISPATCH_CALL' and
+          // 'NPY__CPU_DISPATCH_BASELINE_CALL'.
+          // it highely recomaned to include the config header before exectuing
+        // the dispatching macros in case if there's another header in the scope.
+          #include "hello.dispatch.h"
```
[the remark this answers is not in the document]
Comment (seiko2plus) [reply doc/source/reference/simd/simd-optimizations.rst:469]: Yes, it's safe and clean to add it inside any level of scope, e.g. functions.
This header only contains two abstract C macros that mainly used for determining the required optimizations from outside the dispatch-able sources.

see also [npy_cpu_dispatch.h](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/npy_cpu_dispatch.h) and

https://github.com/numpy/numpy/blob/72b05c04442b3c2dbde5f30795a07c2e60908830/numpy/core/code_generators/generate_umath.py#L1052-L1061

File: doc/source/reference/simd/simd-optimizations.rst [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+      #define DISPATCH_CALL_BASELINE_HIGH_CB(FN, ARGS) \
+        else { FN NPY_EXPAND(ARGS); }
+
+      // NumPy has a macro called 'NPY_CPU_DISPATCH_DECLARE' can be used
+      // for forward declrations any kind of prototypes based on
+      // 'NPY__CPU_DISPATCH_CALL' and 'NPY__CPU_DISPATCH_BASELINE_CALL'.
+      // However in this example, we just handle it manually.
+      void simd_whoami(const char *extra_info);
+      void simd_whoami_AVX512F(const char *extra_info);
+      void simd_whoami_SSE41(const char *extra_info);
```
[the remark this answers is not in the document]
Comment (seiko2plus) [reply doc/source/reference/simd/simd-optimizations.rst:460]: yes. what I was trying todo is explaining what happening under the hood, but now I feel like I need to move these examples into an advanced topic. And add examples for how to use the CPU dispatcher via high-level macros in [npy_cpu_dispatch.h](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/npy_cpu_dispatch.h) without giving too many details.

**Comment by seiko2plus** [comment]
@seberg, I would like to merge this patch as-is and then I'm going to release a series pull-requests to improve it. can we just consider it as a seed doc?

**Comment by seberg** [comment]
OK, works for me, you can just use whatever suggestions in any followup. Lets put it in, thanks @seiko2plus.

**Pull request merged**